### PR TITLE
sh architecture esil implementation [WIP] [NOTFORMERGE] 

### PR DESCRIPTION
**Pull request by jasuramme**
I implemented ESIL commands for all instructions I could (left FPU instructions).
Anyway there is still work in progress. I will test it for few months. And I have plans to change some horrible ESIL constructions I made (when I just started, ESIL almost broke my mind). 

Any suggestions will be appreciated.

### CODE REVIEW

File: libr/anal/p/anal_sh.c [hunk 1]
```diff
@@ -269,87 +289,120 @@ static RAnalValue *anal_regrel_jump(RAnal* anal, RAnalOp* op, ut8 reg){
...
 		op->type = R_ANAL_OP_TYPE_UCALL;
 		op->delay = 1;
 		op->dst = anal_regrel_jump (anal, op, GET_TARGET_REG(code));
+        r_strbuf_setf (&op->esil, "pc,pr,=,r%d,4,+,pc,+=",GET_TARGET_REG(code));
 	} else if (IS_BRAF(code)) {
 		/* Unconditional branch to Rn+PC+4, no delay slot */
 		op->type = R_ANAL_OP_TYPE_UJMP;
 		op->dst = anal_regrel_jump (anal, op, GET_TARGET_REG(code));
 		op->eob = true;
+        r_strbuf_setf (&op->esil, "r%d,4,+,pc,+=",GET_TARGET_REG(code));
```
Comment (radare) [inline libr/anal/p/anal_sh.c]: Use tabs not spaces

File: libr/anal/p/anal_sh.c [hunk 2]
```diff
@@ -184,6 +196,15 @@
 /* index of PC reg in regs[] array*/
 #define PC_IDX 16
 
+static ut64 expand16to32(unsigned int num){
+	ut64 ret = num;
+	if ((ret & 0x8000) != 0)
```
Comment (radare) [inline libr/anal/p/anal_sh.c]: Open brace at right. != 0 is implicit

File: libr/anal/p/anal_sh.c [hunk 3]
```diff
@@ -184,6 +196,15 @@
 /* index of PC reg in regs[] array*/
 #define PC_IDX 16
 
+static ut64 expand16to32(unsigned int num){
```
Comment (radare) [inline libr/anal/p/anal_sh.c]: Ut16 as arg and ut32 as ret?

File: libr/anal/p/anal_sh.c [hunk 4]
```diff
@@ -184,6 +196,15 @@
...
+	ut64 ret = num;
+	if ((ret & 0x8000) != 0)
+	{
```
Comment (radare) [inline libr/anal/p/anal_sh.c]: This can be onelinered with the ternary operator

File: libr/anal/p/anal_sh.c [hunk 5]
```diff
@@ -269,87 +289,120 @@ static RAnalValue *anal_regrel_jump(RAnal* anal, RAnalOp* op, ut8 reg){
...
+        r_strbuf_setf (&op->esil, "r%d,r%d,*,macl,=",GET_TARGET_REG(code),GET_SOURCE_REG(code));
 	} else if (IS_SLEEP(code)) {
 		op->type = R_ANAL_OP_TYPE_UNK;
-	} else if (IS_STSMAC(code)) {	//0000nnnn0000101_ sts MAC*,<REG_N>
+    } else if (IS_STSMACH(code)) {	//0000nnnn0000101_ sts MAC*,<REG_N>
 		op->type = R_ANAL_OP_TYPE_MOV;
 		op->dst = anal_fill_ai_rg (anal, GET_TARGET_REG(code));
-	} else if (IS_STCSR1(code)) {	//0000nnnn00010010 stc {sr,gbr,vbr,ssr},<REG_N>
+        r_strbuf_setf (&op->esil, "mach,r%d,=",GET_TARGET_REG(code));
+    } else if (IS_STSLMACL(code)){
```
Comment (radare) [inline libr/anal/p/anal_sh.c]: Space before (

File: libr/anal/p/anal_sh.c [hunk 6]
```diff
@@ -580,52 +793,67 @@ static int first_nibble_is_8(RAnal* anal, RAnalOp* op, ut16 code){
...
 static int movw_pcdisp_reg(RAnal* anal, RAnalOp* op, ut16 code){
-	op->type = R_ANAL_OP_TYPE_LOAD;
-	op->dst = anal_fill_ai_rg (anal, GET_TARGET_REG(code));
-	//op->src[0] = anal_fill_reg_disp_mem(anal,PC_IDX,code&0xFF,WORD_SIZE);	//XXX trash in 2 commits
-	op->src[0] = anal_pcrel_disp_mov (anal, op, code&0xFF, WORD_SIZE);
+    op->type = R_ANAL_OP_TYPE_LOAD;
+    op->dst = anal_fill_ai_rg (anal, GET_TARGET_REG(code));
+    int tmp=(code&0xFF)*2+op->addr+4;
+    op->src[0]=r_anal_value_new ();
+    op->src[0]->base = tmp;
```
Comment (radare) [inline libr/anal/p/anal_sh.c]: Spaces around =

File: libr/anal/p/anal_sh.c [hunk 7]
```diff
@@ -807,15 +1056,29 @@ static int sh_set_reg_profile(RAnal* anal){
...
 
+static int esil_sh_init (RAnalEsil *esil) {
+    if (esil->anal && esil->anal->reg) {        // initial values
+        r_reg_set_value (esil->anal->reg, r_reg_get (esil->anal->reg, "pc", -1), 0x0000);
+        //r_reg_set_value (esil->anal->reg, r_reg_get (esil->anal->reg, "flags", -1), 0x00);
+    }
+    return true;
+}
+
+static int esil_sh_fini (RAnalEsil *esil) {
```
Comment (radare) [inline libr/anal/p/anal_sh.c]: Remove this callback does nothing

**Comment by XVilka** [comment]
Seems you added accidentally unneeded file from `shlr/sdb`

File: libr/anal/p/anal_sh.c (hunk 5 above)
Comment (jasuramme) [reply libr/anal/p/anal_sh.c]: I don't see what you mean. Maybe space after comma? Before GET_TARGET_REG?

Comment (jasuramme) [reply libr/anal/p/anal_sh.c]: Ah, alright. I'll change it in all macros. Will do like IS_STSLMACL (code)

**Comment by XVilka** [comment]
Would be nice if you add regression tests for it too, something like this one for ARM 64 https://github.com/radare/radare2-regressions/blob/master/t.esil/arm-64

**Comment by jasuramme** [comment]
Thanks for idea. I'll put it on, while I will check code.
Unfortunately I'm not so experienced in reverse engineering, so some stuff with radare2 takes long time) Don't wait tests tomorrow. 
Also should I do something with calling conventions? Can you point me on some file or example, please?

**Comment by Maijin** [comment]
For calling convention you can add file here https://github.com/radare/radare2/blob/master/libr/anal/d/ see cc-powerpc-32.sdb.txt for example

File: libr/anal/p/anal_sh.c [hunk 8]
```diff
@@ -807,15 +1056,25 @@ static int sh_set_reg_profile(RAnal* anal){
...
+		"gpr	macl	.32	88	0\n"
+		"gpr	tmp	.32	88	0\n";
 	return r_reg_set_profile_string(anal->reg, p);
 }
 
 static int archinfo(RAnal *anal, int q) {
 	return 2; /* :) */
 }
 
+static int esil_sh_init(RAnalEsil *esil) {
```
Comment (radare) [inline libr/anal/p/anal_sh.c]: initializing the program counter should be task of rcore, not the esil constructor. so i would kill this function too

File: libr/anal/p/anal_sh.c [hunk 9]
```diff
@@ -761,10 +1009,10 @@ static int (*first_nibble_decode[])(RAnal*,RAnalOp*,ut16) = {
  * routines defined in first_nibble_decode table
  */
 static int sh_op(RAnal *anal, RAnalOp *op, ut64 addr, const ut8 *data, int len) {
-	ut8 op_MSB,op_LSB;
+	ut8 op_MSB, op_LSB;
 	int ret;
 	if (!data || len < 2) {
-		return 0;
+			return 0;
```
Comment (radare) [inline libr/anal/p/anal_sh.c]: wrong indent

**Comment by radare** [comment]
remove the init function and i'll merge

File: libr/anal/p/anal_sh.c (hunk 8 above)
Comment (jasuramme) [reply libr/anal/p/anal_sh.c]: I removed function, but i'm just starting to test that. There still might be dozens of bugs

File: libr/anal/p/anal_sh.c (hunk 9 above)
Comment (jasuramme) [reply libr/anal/p/anal_sh.c]: I removed function, but i'm just starting to test that. There still might be dozens of bugs

**Comment by jasuramme** [comment]
I removed function, but i'm just starting to test that. There still might be dozens of bugs

**Comment by radare** [comment]
Thanks!

**Pull request merged**